src/engine/Account-Imap.cpp

```diff
@@ -327,120 +262,115 @@ GncImportMatchMap::find_account_bayes (GList* tokens)
+            }
+        } // for all accounts in tokenInfo
 
-                /* add the account guid and (struct account_probability*)
-                 * to the hash table */
-                g_hash_table_insert(running_probabilities,
-                                    account_c->account_guid, account_p);
+        // Now look for highest probability
+        for (auto &itr : probability_hash)
+        {
+            /* P(AB) = A*B / [A*B + (1-A)*(1-B)]
+             * NOTE: so we only keep track of a running product(A*B*C...)
+             * and product difference ((1-A)(1-B)...) */
+
+            int32_t probability =
+                (itr.second.m_product /
+                (itr.second.m_product + itr.second.m_product_difference))
+                 * PROBABILITY_FACTOR;
+
+            if (probability > max_probability)
+            {
+                max_probability = probability;
+                selected_account_guid = itr.first;
             }
-        } /* for all accounts in tokenInfo */
+        }
 
         /* free the data in tokenInfo */
         for (current_account_token = tokenInfo.accounts; current_account_token;
                 current_account_token = current_account_token->next)
         {
             /* free up each struct account_token_count we allocated */
-            g_free((struct account_token_count*)current_account_token->data);
+            delete((struct account_token_count*)current_account_token->data);
         }
-
-        g_list_free(tokenInfo.accounts); /* free the accounts GList */
+        g_list_free (tokenInfo.accounts); /* free the accounts GList */
     }
 
-    /* build a hash table of account names and their final probabilities
-     * from each entry in the running_probabilties hash table
-     */
-    g_hash_table_foreach(running_probabilities, buildProbabilities,
-                         final_probabilities);
-
-    /* find the highest probabilty and the corresponding account */
-    memset(&account_i, 0, sizeof(struct account_info));
-    g_hash_table_foreach(final_probabilities, highestProbability, &account_i);
-
-    /* free each element of the running_probabilities hash */
-    g_hash_table_foreach(running_probabilities, freeProbabilities, NULL);
-
-    /* free the hash tables */
-    g_hash_table_destroy(running_probabilities);
-    g_hash_table_destroy(final_probabilities);
-
     PINFO("highest P('%s') = '%d'",
-          account_i.account_guid ? account_i.account_guid : "(null)",
-          account_i.probability);
+          selected_account_guid.c_str() ? selected_account_guid.c_str() : "(null)",
+          max_probability);
 
     /* has this probability met our threshold? */
-    if (account_i.probability >= threshold)
+    if (max_probability >= threshold)
     {
         GncGUID *guid;
         Account *account = NULL;
 
         PINFO("Probability has met threshold");
 
-        guid = g_new (GncGUID, 1);
+        guid = new (GncGUID);
 
```

Comment:
No need for guid to go in the free store. Create it on the stack and remember that the declaration will call the default constructor.